### PR TITLE
Feature/print valid profiles

### PR DESCRIPTION
**Pull request by allcloud-jonathan**
To have an easy view how much time is left and if one can run that long running command or has to get fresh credentials first.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +65,7 @@ Clisso has the following commands:
     get         Get temporary credentials for an app
     help        Help about any command
     providers   Manage providers
+    status      show currently valid profiles
```
Comment (johananl) [inline README.md]: Please keep capitalization consistent :-)
Also, it's not clear to me as a user what "valid" means, so I would maybe rephrase as follows:

`Show active (non-expired) credentials`.

File: aws/aws.go [hunk 2]
```diff
@@ -19,6 +19,15 @@ type Credentials struct {
 	Expiration      time.Time
 }
 
+type ProfileStatus struct {
+	Name         string
+	LivetimeLeft time.Duration
```
Comment (johananl) [inline aws/aws.go]: "Livetime" should probably be "Lifetime".

File: aws/aws.go [hunk 3]
```diff
@@ -19,6 +19,15 @@ type Credentials struct {
 	Expiration      time.Time
 }
 
+type ProfileStatus struct {
```
Comment (johananl) [inline aws/aws.go]: If `Profiles` is a slice of `ProfileStatus`-es, I think we should rename `type ProfileStatus` to `Profile`.

File: aws/aws.go [hunk 4]
```diff
@@ -69,3 +78,24 @@ func WriteToShell(c *Credentials, windows bool, w io.Writer) {
 		)
 	}
 }
+
+// GetNonExpiredCredentials Return profiles which have a aws_expiration key but are not yet expired
```
Comment (johananl) [inline aws/aws.go]: Please add a period at the end :-)

This function could return profiles which weren't created by Clisso... I'm thinking whether we should prefix all profiles Clisso saves with `clisso-`. This would also make it clearer we're using temporary credentials when doing `export AWS_PROFILE=profile`. WDYT?

File: aws/aws.go [hunk 5]
```diff
@@ -69,3 +78,24 @@ func WriteToShell(c *Credentials, windows bool, w io.Writer) {
...
+func GetNonExpiredCredentials(filename string) (Profiles, error) {
```
Comment (johananl) [inline aws/aws.go]: I would return a *pointer* to `Profiles`, that is:

`func GetNonExpiredCredentials(filename string) (*Profiles, error) {}`

And then return `nil` on error. It's more common in Golang to do it this way, and it makes it clearer there was a problem since we return `nil` and not an empty slice.

File: aws/aws.go [hunk 6]
```diff
@@ -69,3 +78,24 @@ func WriteToShell(c *Credentials, windows bool, w io.Writer) {
...
+	profiles := Profiles{}
+	cfg, err := ini.LooseLoad(filename)
+	if err != nil {
+		return profiles, err
+	}
+	for _, s := range cfg.Sections() {
+		if s.HasKey("aws_expiration") {
+			v, err := s.Key("aws_expiration").TimeFormat(time.RFC3339)
+			if err == nil {
```
Comment (johananl) [inline aws/aws.go]: The Golang idiom here is to check if we **do** have an error. If we do, we return. If not, we keep going.
This removes one layer of nesting and makes the code more readable. So:

```
if err != nil {
    return nil, err
}

if time.Now()...
```

File: aws/aws_test.go [hunk 7]
```diff
@@ -24,22 +24,54 @@ func TestWriteToFile(t *testing.T) {
...
-	p := "testprofile"
+	p := "expiredprofile"
 
 	// Write credentials
 	err := WriteToFile(&c, fn, p)
 	if err != nil {
 		t.Fatal("Could not write credentials to file: ", err)
 	}
 
+	// sleep so above credentials expire
```
Comment (johananl) [inline aws/aws_test.go]: Cap it please ;-)

Why not write an expired timestamp to the file instead of writing `Now()` and sleeping? This delays **every** run of the unit tests by 2 seconds...

File: aws/aws_test.go [hunk 8]
```diff
@@ -11,9 +11,9 @@ import (
 )
 
 func TestWriteToFile(t *testing.T) {
```
Comment (johananl) [inline aws/aws_test.go]: I would move the expiration test to a new test case. A test case should ideally test one thing.
The `TestWriteToFile` test case should only check that writing credentials to file works. We should have *another* test case which checks that expiration logic is handled properly. Doing so will make it easier to figure out what is wrong upon test failure.

File: aws/aws_test.go [hunk 9]
```diff
@@ -24,22 +24,54 @@ func TestWriteToFile(t *testing.T) {
...
 	if err != nil {
 		t.Fatal("Could not load INI file: ", err)
 	}
 
+	p = "expiredprofile"
 	s := cfg.Section(p)
 
+	// Verify File
+	if s.HasKey("aws_access_key_id") || s.HasKey("aws_secret_access_key") || s.HasKey("aws_session_token") || s.HasKey("aws_expiration") {
+		t.Fatal("expired profile was not cleaned up")
```
Comment (johananl) [inline aws/aws_test.go]: Cap it please ;-)

File: aws/aws_test.go [hunk 10]
```diff
@@ -80,6 +112,92 @@ func TestWriteToFile(t *testing.T) {
...
+	exp := time.Now()
+
+	c := Credentials{
+		AccessKeyID:     id,
+		SecretAccessKey: sec,
+		SessionToken:    tok,
+		Expiration:      exp,
+	}
+
+	// aboves credentials are expiring this excact moment..
```
Comment (johananl) [inline aws/aws_test.go]: Cap cap cap ;-)

Also, there are many typos here :-/

File: aws/aws_test.go [hunk 11]
```diff
@@ -80,6 +112,92 @@ func TestWriteToFile(t *testing.T) {
...
+	p := "expired"
+
+	// Write credentials
+	err := WriteToFile(&c, fn, p)
+	if err != nil {
+		t.Fatal("Could not write credentials to file: ", err)
+	}
+
+	// expire in 1 hour
```
Comment (johananl) [inline aws/aws_test.go]: Cap it cap'n ;-)

File: aws/aws_test.go [hunk 12]
```diff
@@ -80,6 +112,92 @@ func TestWriteToFile(t *testing.T) {
...
+	c.Expiration = time.Now().Add(time.Duration(1) * time.Hour)
+	p = "valid"
+	// Write credentials
```
Comment (johananl) [inline aws/aws_test.go]: Missing blank line?

File: aws/aws_test.go [hunk 13]
```diff
@@ -80,6 +112,92 @@ func TestWriteToFile(t *testing.T) {
...
+	}
+
+	// Read credentials
+	cfg, err := ini.Load(fn)
+	if err != nil {
+		t.Fatal("Could not load INI file: ", err)
+	}
+
+	s := cfg.Section("not-there")
+	// Verify File
```
Comment (johananl) [inline aws/aws_test.go]: Missing blank line?

File: cmd/get.go [hunk 14]
```diff
@@ -107,5 +107,6 @@ If no app is specified, the selected app (if configured) will be assumed.`,
 		} else {
 			log.Fatalf(color.RedString("Unsupported identity provider type '%s' for app '%s'"), pType, app)
 		}
+		outputStatus()
```
Comment (johananl) [inline cmd/get.go]: I would rename to `printStatus()`.

File: cmd/status.go [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+var cmdStatus = &cobra.Command{
+	Use:   "status",
+	Short: "show currently valid profiles",
+	Long:  `show currently valid profiles`,
+	Run: func(cmd *cobra.Command, args []string) {
+		outputStatus()
+	},
+}
+
+func outputStatus() {
```
Comment (johananl) [inline cmd/status.go]: I would rename to `printStatus()`.

File: cmd/status.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/allcloud-io/clisso/aws"
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func init() {
+	RootCmd.AddCommand(cmdStatus)
+	cmdStatus.Flags().StringVarP(
+		&writeToFile, "write-to-file", "w", "",
```
Comment (johananl) [inline cmd/status.go]: Why does `clisso status` need the `-w` flag? This command should only show stuff to the user, not do anything with the credentials, right?

File: cmd/status.go [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+		"Write credentials to this file instead of the default ($HOME/.aws/credentials)",
+	)
+	viper.BindPFlag("global.credentials-path", cmdStatus.Flags().Lookup("write-to-file"))
+}
+
+var cmdStatus = &cobra.Command{
+	Use:   "status",
+	Short: "show currently valid profiles",
```
Comment (johananl) [inline cmd/status.go]: Please keep the capitalization consistent... It looks bad if some commands look different:

```
Usage:
  clisso [command]

Available Commands:
  apps        Manage apps
  get         Get temporary credentials for an app
  help        Help about any command
  providers   Manage providers
  status      show currently valid profiles
  version     Show version info
```

**Comment by allcloud-jonathan** [comment]
* About the prefix: Definitely a no. Because that imposes how people have to code / name their profiles. It makes it hard to switch from static keys to temporary ones.
We're so far the only solution to write the expiry field. If it matches the output of another program, yes it will be cleaned. But I've used it in conjunction with static keys and an ADFS based solution without issues.
* About writing expired credentials directly to the file instead of sleeping. The way the code is constructed, these will never be written to file. They will be immediately cleaned up before write in the block at [aws/aws.go line36](https://github.com/allcloud-io/clisso/blob/master/aws/aws.go#L36)

**Comment by allcloud-jonathan** [comment]
@johananl Any comment?

**Comment by johananl** [comment]
> @johananl Any comment?

Will look again ASAP.

**Comment by johananl** [comment]
Rebased over master.

File: cmd/status.go [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+func printStatus() {
+	configfile, err := homedir.Expand(viper.GetString("global.credentials-path"))
+	profiles, err := aws.GetNonExpiredCredentials(configfile)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Print("The following profiles are currently not expired:")
+	for _, p := range profiles.Profiles {
+		log.Printf("%v: remaining time %v", p.Name, p.LifetimeLeft.Round(time.Second))
```
Comment (johananl) [inline cmd/status.go]: How about printing this info in a nice table?

We can maybe use this library: https://github.com/olekukonko/tablewriter

I would include the following columns:

- App (string)
- Expires at (timestamp)
- Remaining time (time delta)

File: aws/aws.go [hunk 19]
```diff
@@ -19,6 +19,15 @@ type Credentials struct {
 	Expiration      time.Time
 }
 
+type Profile struct {
+	Name         string
+	LifetimeLeft time.Duration
+}
+
+type Profiles struct {
```
Comment (johananl) [inline aws/aws.go]: Looks like the only field in `Profiles` is a slice of `Profile`s. In this case there is no need to create another custom type, we can simply create a slice of `Profile`s instead of initializing a `Profiles` struct:

```go
profiles := []Profile{}

// Add stuff to slice.
for _, x := range whatever {
    if stuff {
        profiles := append(profiles, x)
    }
}
```

The `Profiles` type only adds unnecessary wrapping in this case.

File: aws/aws_test.go (hunk 8 above)
Comment (johananl) [reply aws/aws_test.go]: How about it?

File: aws/aws_test.go [hunk 20]
```diff
@@ -24,22 +24,54 @@ func TestWriteToFile(t *testing.T) {
...
+	p := "expiredprofile"
 
 	// Write credentials
 	err := WriteToFile(&c, fn, p)
 	if err != nil {
 		t.Fatal("Could not write credentials to file: ", err)
 	}
 
+	// Sleep so above credentials expire
+	time.Sleep(time.Duration(2) * time.Second)
```
[the remark this answers is not in the document]
Comment (johananl) [reply aws/aws_test.go]: Hello bot. I strongly agree. I'm sure we can find a way to simulate the expiry without sleeping.

File: aws/aws_test.go [hunk 21]
```diff
@@ -24,22 +24,54 @@ func TestWriteToFile(t *testing.T) {
...
 	cfg, err := ini.Load(fn)
 	if err != nil {
 		t.Fatal("Could not load INI file: ", err)
 	}
 
+	p = "expiredprofile"
 	s := cfg.Section(p)
 
+	// Verify File
+	if s.HasKey("aws_access_key_id") || s.HasKey("aws_secret_access_key") || s.HasKey("aws_session_token") || s.HasKey("aws_expiration") {
```
Comment (johananl) [inline aws/aws_test.go]: A loop could be more concise here:

```go
keys := []string{"foo", "bar"}
for _, k := range keys {
    if s.HasKey(k) {
        t.Fatal("Oh no!")
    }
}
```
